File: python/microsoft-agents-hosting/microsoft/agents/hosting/core/agent_context.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""Agent context implementation."""

import logging
from typing import Any, Optional

from .types import IAgentContext, IThreadState, IStateStore, CancellationToken


class ThreadState(IThreadState):
    """Thread state implementation."""

    def __init__(self, thread_id: str, state_store: IStateStore):
        self._thread_id = thread_id
        self._state_store = state_store

    async def get_async(self, key: str, default: Any = None) -> Any:
        """Get a state value."""
        value = await self._state_store.get_async(self._thread_id, key)
        return value if value is not None else default

    async def set_async(self, key: str, value: Any) -> None:
        """Set a state value."""
        await self._state_store.set_async(self._thread_id, key, value)

    async def delete_async(self, key: str) -> None:
        """Delete a state value."""
        await self._state_store.delete_async(self._thread_id, key)

    async def get_all_async(self) -> dict[str, Any]:
        """Get all state."""
        return await self._state_store.get_all_async(self._thread_id)

    async def clear_async(self) -> None:
        """Clear all state."""
        await self._state_store.clear_async(self._thread_id)


class AgentContext(IAgentContext):
    """Agent context implementation."""

    def __init__(
        self,
        run_id: str,
        thread_id: str,
        state_store: IStateStore,
        logger: logging.Logger,
        response_callback: Any = None,
    ):
        self._run_id = run_id
        self._thread_id = thread_id
        self._state = ThreadState(thread_id, state_store)
        self._logger = logger
        self._response_callback = response_callback
        self._responses: list[str] = []

    @property
    def run_id(self) -> str:
        """Gets the current run ID."""
        return self._run_id

    @property
    def thread_id(self) -> str:
        """Gets the current thread ID."""
        return self._thread_id

    @property
    def state(self) -> IThreadState:
        """Gets the state manager for this thread."""
        return self._state

    async def respond_async(
        self, content: str, cancellation_token: Optional[CancellationToken] = None
    ) -> None:
        """Send a response message to the user."""
        if cancellation_token and cancellation_token.is_cancelled():
            return

        self._responses.append(content)
        if self._response_callback:
            await self._response_callback(content)

    async def log_async(
        self,
        message: str,
        level: str = "INFO",
        cancellation_token: Optional[CancellationToken] = None,
    ) -> None:
        """Log a message for debugging/observability."""
        if cancellation_token and cancellation_token.is_cancelled():
            return

        log_level = getattr(logging, level.upper(), logging.INFO)
        self._logger.log(log_level, message, extra={"thread_id": self._thread_id, "run_id": self._run_id})

    async def pause_for_approval_async(
        self,
        summary: str,
        metadata: Optional[dict[str, Any]] = None,
        cancellation_token: Optional[CancellationToken] = None,
    ) -> None:
        """Pause the run and wait for approval before continuing."""
        if cancellation_token and cancellation_token.is_cancelled():
            return

        # TODO: Implement pause mechanism
        await self.log_async(
            f"Pausing for approval: {summary}",
            level="INFO",
            cancellation_token=cancellation_token,
        )

File: dotnet/src/Microsoft.Agents.Protocol.Hosting/Core/AgentContext.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Agents;
using Microsoft.Extensions.Logging;

namespace Microsoft.Agents.Protocol.Hosting.Core;

/// <summary>
/// Implementation of IAgentContext for agent turn processing.
/// </summary>
public class AgentContext : IAgentContext
{
    private readonly ILogger _logger;
    private readonly Func<string, Task>? _responseCallback;

    /// <summary>
    /// Creates a new AgentContext instance.
    /// </summary>
    /// <param name="runId">The run ID.</param>
    /// <param name="threadId">The thread ID.</param>
    /// <param name="storage">The storage implementation.</param>
    /// <param name="logger">The logger.</param>
    /// <param name="responseCallback">Optional callback for responses.</param>
    public AgentContext(
        string runId,
        string threadId,
        IStorage storage,
        ILogger logger,
        Func<string, Task>? responseCallback = null)
    {
        RunId = runId ?? throw new ArgumentNullException(nameof(runId));
        ThreadId = threadId ?? throw new ArgumentNullException(nameof(threadId));
        State = new AgentState(threadId, storage ?? throw new ArgumentNullException(nameof(storage)));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _responseCallback = responseCallback;
    }

    /// <inheritdoc/>
    public string RunId { get; }

    /// <inheritdoc/>
    public string ThreadId { get; }

    /// <inheritdoc/>
    public ILogger Logger => _logger;

    /// <summary>
    /// Gets the state storage for this context.
    /// </summary>
    public AgentState State { get; }

    /// <inheritdoc/>
    public async Task RespondAsync(AIContent content, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return;
        }

        if (_responseCallback != null)
        {
            // Convert AIContent to text representation
            var text = content switch
            {
                TextContent textContent => textContent.Text ?? string.Empty,
                _ => content?.ToString() ?? string.Empty
            };
            await _responseCallback(text);
        }
    }

    /// <inheritdoc/>
    public async Task RespondAsync(string text, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return;
        }

        if (_responseCallback != null)
        {
            await _responseCallback(text);
        }
    }

    /// <inheritdoc/>
    public Task LogAsync(string message, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.CompletedTask;
        }

        _logger.LogInformation("[Run {RunId}] {Message}", RunId, message);
        return Task.CompletedTask;
    }

    /// <summary>
    /// Logs a message with a specific level.
    /// </summary>
    /// <param name="message">The message to log.</param>
    /// <param name="level">The log level (DEBUG, INFO, WARNING, ERROR).</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public Task LogAsync(string message, string level, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.CompletedTask;
        }

        var logLevel = level?.ToUpperInvariant() switch
        {
            "DEBUG" => LogLevel.Debug,
            "WARNING" => LogLevel.Warning,
            "ERROR" => LogLevel.Error,
            _ => LogLevel.Information
        };

        _logger.Log(logLevel, "[Run {RunId}] {Message}", RunId, message);
        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    public Task PauseForApprovalAsync(
        string summary,
        object? metadata = null,
        CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.CompletedTask;
        }

        _logger.LogInformation("[Run {RunId}] Pause for approval requested: {Summary}", RunId, summary);

        // In a real implementation, this would pause the run and wait for external approval
        // For now, we just log it

        return Task.CompletedTask;
    }
}
